packages/test/src/test-custom-payload-codec.ts

```diff
@@ -0,0 +1,181 @@
+          fn(_, message) {
+            logs.push(message);
+          },
+        },
+      },
+    };
+
+    const dataConverter = { payloadCodec: new TestEncodeCodec() };
+    const taskQueue = 'test-workflow-encoded';
+    const worker = await Worker.create({
+      ...defaultOptions,
+      taskQueue,
+      dataConverter, // TODO { payloadCodec: defaultPayloadCodec }
+      sinks,
+    });
+    const client = new WorkflowClient(new Connection().service, { dataConverter });
+    const runAndShutdown = async () => {
+      const result = await client.execute(twoStrings, {
+        args: ['arg1', 'arg2'],
+        workflowId: uuid4(),
+        taskQueue,
+      });
+
+      t.is(result, 'encoded');
+      worker.shutdown();
+    };
+    await Promise.all([worker.run(), runAndShutdown()]);
+    t.is(logs[0], 'encodedencoded');
+  });
+
+  test('Workflow arguments and retvals are decoded', async (t) => {
+    const logs: string[] = [];
+    const sinks: InjectedSinks<LogSinks> = {
+      logger: {
+        log: {
+          fn(_, message) {
+            logs.push(message);
+          },
+        },
+      },
+    };
+
+    const dataConverter = { payloadCodec: new TestDecodeCodec() };
+    const taskQueue = 'test-workflow-decoded';
+    const worker = await Worker.create({
+      ...defaultOptions,
+      taskQueue,
+      dataConverter, // TODO { payloadCodec: defaultPayloadCodec }
+      sinks,
+    });
+    const client = new WorkflowClient(new Connection().service, { dataConverter });
+    const runAndShutdown = async () => {
+      const result = await client.execute(twoStrings, {
+        args: ['arg1', 'arg2'],
+        workflowId: uuid4(),
+        taskQueue,
+      });
+
+      t.is(result, 'decoded');
+      worker.shutdown();
+    };
+    await Promise.all([worker.run(), runAndShutdown()]);
+    t.is(logs[0], 'decodeddecoded');
+  });
+
+  test('Activity arguments and retvals are encoded', async (t) => {
+    const workflowLogs: string[] = [];
+    const sinks: InjectedSinks<LogSinks> = {
+      logger: {
+        log: {
+          fn(_, message) {
+            workflowLogs.push(message);
+          },
+        },
+      },
+    };
+    const activityLogs: string[] = [];
+
+    const dataConverter = { payloadCodec: new TestEncodeCodec() };
+    const taskQueue = 'test-activity-encoded';
+    const worker = await Worker.create({
+      ...defaultOptions,
+      activities: createConcatActivity(activityLogs),
+      taskQueue,
+      dataConverter,
+      sinks,
+    });
+    const client = new WorkflowClient(new Connection().service, { dataConverter });
+    const runAndShutdown = async () => {
+      const result = await client.execute(twoStringsActivity, {
+        workflowId: uuid4(),
+        taskQueue,
+      });
+
+      t.is(result, 'encoded');
+      worker.shutdown();
+    };
+    await Promise.all([worker.run(), runAndShutdown()]);
+    t.is(workflowLogs[0], 'encoded');
+    t.is(activityLogs[0], 'Activityencodedencoded');
```

Comment:
nit: You can probably test this without the logger if you use a single codec for both encoding and decoding.